sycl/source/detail/graph_impl.cpp

```diff
@@ -77,29 +77,198 @@ void duplicateNode(const std::shared_ptr<node_impl> Node,
 /// Recursively add nodes to execution stack.
 /// @param NodeImpl Node to schedule.
 /// @param Schedule Execution ordering to add node to.
+/// @param PartitionBounded If set to true, the topological sort is stopped at
+/// partition borders. Hence, nodes belonging to a partition different from the
+/// NodeImpl partition are not processed.
 void sortTopological(std::shared_ptr<node_impl> NodeImpl,
-                     std::list<std::shared_ptr<node_impl>> &Schedule) {
+                     std::list<std::shared_ptr<node_impl>> &Schedule,
+                     bool PartitionBounded = false) {
   for (auto &Succ : NodeImpl->MSuccessors) {
-    // Check if we've already scheduled this node
     auto NextNode = Succ.lock();
+    if (PartitionBounded &&
+        (NextNode->MPartitionNum != NodeImpl->MPartitionNum)) {
+      continue;
+    }
+    // Check if we've already scheduled this node
     if (std::find(Schedule.begin(), Schedule.end(), NextNode) ==
         Schedule.end()) {
-      sortTopological(NextNode, Schedule);
+      sortTopological(NextNode, Schedule, PartitionBounded);
     }
   }
 
   Schedule.push_front(NodeImpl);
 }
+
+/// Propagates the partition number `PartitionNum` to predecessors
```

Comment:
```suggestion
/// Propagates the partition number `PartitionNum` to predecessors.
```

Nitpicking, but overall this comment is very helpful👍